### PR TITLE
Add AMRules

### PR DESCRIPTION
**Pull request by smastelini**
AMRules is a classical rule-based regressor for data stream mining. It's described in this [paper](https://dl.acm.org/doi/abs/10.1145/2829955).

**TODO**
- [x] Documentation
- [x] Extensive testing

### CODE REVIEW

**Comment by smastelini** [comment]
I decided to apply the same nomenclature used in the original paper.

That's why I use, for instance, `n_min` rather than `grace_period`. I think it's easier to match the code to the paper this way. Although the `rule` module is related to `tree`, I believe it's alright to use a somewhat different naming convention in rule-based algorithms.

I am more familiar with regression algorithms, but I might give Very Fast Decision Rules (the classification counterpart of AMRules) a try in the future.

**Comment by smastelini** [comment]
The bug fix in [54ee264](https://github.com/online-ml/river/pull/594/commits/54ee2648be4fb63f7a67b521aa9c0089481983d9) was possible after a careful inspection of MOA's code (the paper doesn't describe this specific step). It was a silly mistake that ended up making AMRules' predictions a little bit worse. I left a comment in the code for future reference.

File: river/rule/amrules.py [hunk 1]
```diff
@@ -95,6 +96,25 @@ def drift_test(self, y, y_pred):
         return in_drift
 
     def anomaly_score(self, x):
```
Comment (MaxHalford) [inline river/rule/amrules.py:98]: We call this `score_one` in the `AnomalyDetector` base class :)

File: river/rule/amrules.py [hunk 2]
```diff
@@ -95,6 +96,25 @@ def drift_test(self, y, y_pred):
...
+        """Rule anomaly score.
+
+        The more negative the score, the more anomalous is the instance regarding the subspace
+        covered by the rule. Instances whose score is greater than zero are considered normal.
+        Different threshold values can be used to decide upon discarding or using the instance
+        for training. A small negative threshold is considered to be a conservative choice,
+        whereas cutting at zero is a more aggressive choice.
```
Comment (MaxHalford) [inline river/rule/amrules.py:105]: If you inherit from `AnomalyDetector`, you can just use the description from there to explain this method, no need to repeat ourselves. Your explanation is more verbose so don't hesitate to merge both comments into something better.

File: river/rule/amrules.py [hunk 3]
```diff
@@ -215,8 +298,7 @@ def _new_rule(self) -> RegRule:
             predictor = copy.deepcopy(self.pred_model)
         else:  # adaptive predictor
             predictor = AdaptiveRegressor(
-                model_predictor=copy.deepcopy(self.pred_model),
-                alpha=self.model_selector_decay,
+                model_predictor=copy.deepcopy(self.pred_model), alpha=self.alpha,
```
Comment (MaxHalford) [inline river/rule/amrules.py:326]: You can also use the `.clone()` method that each model now has I believe.

File: river/__init__.py [hunk 4]
```diff
@@ -34,6 +34,7 @@
     preprocessing,
     proba,
     reco,
+    rule,
```
[the remark this answers is not in the document]
Comment (smastelini) [reply river/__init__.py:37]: Good call! I like that :D

I'll change the module name.

File: river/rule/amrules.py [hunk 5]
```diff
@@ -0,0 +1,354 @@
...
+import copy
+import functools
+import io
+import math
+import typing
+
+from river import base, drift, linear_model, stats, tree
+
+from ..tree.split_criterion import VarianceRatioSplitCriterion
+from ..tree.splitter.nominal_splitter_reg import NominalSplitterReg
```
[the remark this answers is not in the document]
Comment (smastelini) [reply river/rule/amrules.py]: :D

Trees and rule-based models are closely related!

[the remark this answers is not in the document]
Comment (smastelini) [reply river/rule/amrules.py]: We intend to also reuse some tree submodules for the rule-based classifiers.

File: river/rule/amrules.py [hunk 6]
```diff
@@ -0,0 +1,354 @@
...
+from .base import HoeffdingRule
+
+
+class MeanRegressor(base.Regressor):
```
[the remark this answers is not in the document]
Comment (smastelini) [reply river/rule/amrules.py:16]: I initially did that, but the dummy regressor does not support passing the `w` parameter to `stats.Mean`. I could create a small parallel PR to extend the `StatisticRegressor` to receive `**kwargs`. What do you think?

[the remark this answers is not in the document]
Comment (smastelini) [reply river/rule/amrules.py:16]: Or use a loop :/ (but it sounds like a waste)

File: river/rule/amrules.py (hunk 2 above)
Comment (smastelini) [reply river/rule/amrules.py:105]: Thanks for the tip! I did as you said. I kept most from the old description because the Rules' scores have a different interpretation than the base anomaly detector. Negative values indicate anomalies, whereas the positive ones indicate normal examples.

File: river/rule/amrules.py (hunk 1 above)
Comment (smastelini) [reply river/rule/amrules.py:98]: I changed that! @MaxHalford, what about the main AMRules class? It does not return a single value, but a tuple. Should I also inherit from `base.AnomalyDetector`?

File: river/rule/amrules.py (hunk 3 above)
Comment (smastelini) [reply river/rule/amrules.py]: Good call! Just changed that in [8d249f4](https://github.com/online-ml/river/pull/594/commits/8d249f4dba1471bdf054af1a31a32a2ca927811c)

File: river/rules/base.py [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+        return False
+
+    def describe(self):
+        if not self.neg:
+            return f"{self.on} = {self.at}"
+        else:
+            return f"{self.on} ≠ {self.at}"
+
+
+class HoeffdingRule(base.Estimator, metaclass=abc.ABCMeta):
```
Comment (jacobmontiel) [inline river/rules/base.py:67]: Please provide a short description.

File: river/rules/base.py [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+        for att_id, splitter in self.splitters.items():
+            suggestions.append(
+                splitter.best_evaluated_split_suggestion(
+                    criterion=self.split_criterion,
+                    pre_split_dist=self.statistics,
+                    att_idx=att_id,
+                    binary_only=True,
+                )
+            )
+        suggestions.sort()
```
Comment (jacobmontiel) [inline river/rules/base.py:164]: What is the criterium for this sorting?

File: river/rules/base.py [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+    def _update_target_stats(self, y, w):
+        pass
+
+    def _update_feature_stats(self, feat_name, feat_val, w):
+        # By default, no feature stats are kept
+        pass
+
+    def update(self, x, y, w):
+        self._total_weight += w
+        self._update_target_stats(y, w)
```
Comment (jacobmontiel) [inline river/rules/base.py:259]: is this necessary?

**Review by jacobmontiel — COMMENTED** [review]
Hi @smastelini 

The PR is in very good shape. Just left a couple of minor comments and questions.

File: river/rules/base.py (hunk 8 above)
Comment (smastelini) [reply river/rules/base.py:164]: The split merit. The `BranchFactory` can be sorted :)

https://github.com/online-ml/river/blob/eda2723b10ae22738bdc063fca3846e4ff601597/river/tree/utils.py#L99-L103

File: river/rules/amrules.py [hunk 10]
```diff
@@ -0,0 +1,526 @@
...
+    def learn_one(self, x: dict, y: base.typing.RegTarget, w: int = 1):
+        self.mean.update(y, w)
+        return self
+
+    def predict_one(self, x: dict):
+        return self.mean.get()
+
+
+class AdaptiveRegressor(base.Regressor):
+    def __init__(self, model_predictor: base.Regressor, alpha: float):
```
[the remark this answers is not in the document]
Comment (smastelini) [reply river/rules/amrules.py:33]: Description added in commit [6978b60](https://github.com/online-ml/river/pull/594/commits/6978b6054b0b5b15dc86dd373bcb4f4f48b5ac6b)!

File: river/rules/amrules.py [hunk 11]
```diff
@@ -0,0 +1,526 @@
...
+        return self
+
+    def predict_one(self, x: dict):
+        return self.pred_model.predict_one(x)
+
+
+class AMRules(base.Regressor):
+    """Adaptive Model Rules.
+
+    AMRules[^1] is rule-based algorithm for incremental regression tasks. AMRules relies on the
```
[the remark this answers is not in the document]
Comment (smastelini) [reply river/rules/amrules.py]: Fixed in commit [6978b60](https://github.com/online-ml/river/pull/594/commits/6978b6054b0b5b15dc86dd373bcb4f4f48b5ac6b)! Thanks for the suggestion!

File: river/rules/amrules.py [hunk 12]
```diff
@@ -0,0 +1,526 @@
...
+    additional capacities not usually found in decision trees.
+
+    Firstly, each created decision rule has a built-in drift detection mechanism. Every time a
+    drift is detected, the affected decision rule is removed. In addition, AMRules' rules also
+    have anomaly detection capabilities. After a warm-up period, each rule tests whether or not
+    the incoming instances are anomalies. Anomalous instances are not used for training.
+
+    Every time no rule is covering an incoming example, a default rule is used to learn
+    from it. A rule covers an instance when all of the rule's literals (tests joined by the
+    logical operation`and`) match the input case. The default rule is also applied for predicting
```
[the remark this answers is not in the document]
Comment (smastelini) [reply river/rules/amrules.py]: Fixed in commit [6978b60](https://github.com/online-ml/river/pull/594/commits/6978b6054b0b5b15dc86dd373bcb4f4f48b5ac6b)!

File: river/rules/amrules.py [hunk 13]
```diff
@@ -0,0 +1,526 @@
...
+
+    def learn_one(self, x: dict, y: base.typing.RegTarget, w: int = 1) -> "AMRules":
+        any_covered = False
+        to_del = set()
+
+        for rule_id, rule in self._rules.items():
+            if not rule.covers(x):
+                continue
+
+            if rule.total_weight > self.m_min:
```
[the remark this answers is not in the document]
Comment (smastelini) [reply river/rules/amrules.py]: Good catch! Fixed in commit [6978b60](https://github.com/online-ml/river/pull/594/commits/6978b6054b0b5b15dc86dd373bcb4f4f48b5ac6b).

File: river/rules/amrules.py [hunk 14]
```diff
@@ -0,0 +1,526 @@
...
+        abs_error_model = abs(y - self.model_predictor.predict_one(x))  # noqa
+
+        self._mae_mean = self.alpha * self._mae_mean + abs_error_mean
+        self._mae_model = self.alpha * self._mae_model + abs_error_model
+
+        self.mean_predictor.learn_one(x, y, w)
+
+        try:
+            self.model_predictor.learn_one(x, y, w)  # noqa
+        except TypeError:
```
[the remark this answers is not in the document]
Comment (smastelini) [reply river/rules/amrules.py:52]: The case where the chosen learning models do not support weights. In these cases, we do a for loop to ensure compatibility with online bagging-based ensembles.

[the remark this answers is not in the document]
Comment (smastelini) [reply river/rules/amrules.py:52]: It's not ideal. I didn't find a totally effective way of checking whether or not the regressors support weights. The solution used in some of the trees is a bit hacky for my taste.

 Ideally, all learners should have support to frequency-based weights, i.e., basically integers, as used in online bagging.

File: river/rules/amrules.py [hunk 15]
```diff
@@ -0,0 +1,526 @@
...
+        hits = 0
+
+        for feat_name, feat_val in x.items():
+            # Skip nominal features
+            if feat_name in self.nominal_features:
+                continue
+
+            mean = self._feat_stats[feat_name].mean.get()
+            var = self._feat_stats[feat_name].get()
+            std = math.sqrt(var)
```
[the remark this answers is not in the document]
Comment (smastelini) [reply river/rules/amrules.py]: Good catch. Indeed, there's no need to compute the `std`. I simplified the code in commit [2745f34](https://github.com/online-ml/river/pull/594/commits/2745f3495b9dc50ecb30f62be4230fd05f17499d).

File: river/rules/amrules.py [hunk 16]
```diff
@@ -0,0 +1,526 @@
...
+        rules covering that example.
+    min_samples_split
+        The minimum number of samples each partition of a binary split candidate must have
+        to be considered valid.
+
+    Notes
+    -----
+    AMRules treats all the non-numerical inputs as nominal features. All instances of
+    `numbers.Number` will be treated as continuous, even if they represent integer categories.
+    When using nominal features, `pred_type` should be set to "mean", otherwise errors will be
```
[the remark this answers is not in the document]
Comment (smastelini) [reply river/rules/amrules.py:225]: Thanks for the observation. I added the missing section in commit [2745f34](https://github.com/online-ml/river/pull/594/commits/2745f3495b9dc50ecb30f62be4230fd05f17499d).

File: river/rules/base.py (hunk 7 above)
Comment (smastelini) [reply river/rules/base.py:67]: I added some additional descriptions in commit [2745f34](https://github.com/online-ml/river/pull/594/commits/2745f3495b9dc50ecb30f62be4230fd05f17499d).

File: river/rules/base.py (hunk 9 above)
Comment (smastelini) [reply river/rules/base.py:283]: If I'm not wrong, I had to keep a separate estimate of the observed weight because the target (split) stats are reset after each expansion. Hence, `total_weight` cannot be obtained reliably from the target stats.

Although the newly expanded rules are not ready to be expanded again yet, they can perform anomaly detection.

AMRules' rules have a warm start before being used to detect anomalies. `total_weight` is checked for this end.

File: river/rules/amrules.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (jacobmontiel) [reply river/rules/amrules.py:52]: I understand. We have the same issue in SRP. Thanks for the clarification.

**Comment by smastelini** [comment]
Time to benchmark!

Here's what I did: I compared River's AMRules against the MOA code (release 2019.05.0).

*Disclaimer:* I'm not sure the memory measurements in MOA are working as intended.

**Datasets** (100k instances, using `seed=42`)

- FriedmanDrift ("lea"): `position=(25000, 50000, 75000)`
- FriedmanDrift ("gra"): `position=(35000, 75000)`
- FriedmanDrift ("gsg"): `position=(35000, 75000), transition_window=2000`

**Hyperparameters**

- Almost all of them kept at their default values
- River's AMRules using `splitter=tree.splitter.TEBSTSplitter(digits=1)`

---

## Results 

friedman_lea
friedman_gra
friedman_gsg

---

## River's AMRules using ADWIN as drift detector

friedman_lea
friedman_gra
friedman_lea

File: river/rules/base.py (hunk 9 above)
Comment (jacobmontiel) [reply river/rules/base.py:283]: Sorry, I was not clear. `self._update_target_stats` is not implemented. Is it implemented somewhere else?

Comment (jacobmontiel) [reply river/rules/base.py:283]: Sorry, I was not clear. `self._update_target_stats` is not implemented. Is it implemented somewhere else?

**Review by jacobmontiel — APPROVED** [review]
You read my mind @smastelini! I was thinking precisely about the evaluation.

The memory calculation in MOA is a little tricky, as it is an estimation of memory consumption over time. Although the fact that it remains constant is strange indeed.

On the other hand, quite interesting results on time. I was not expecting that.

I am happy with the implementation and the tests. I am approving the PR.

**Comment by smastelini** [comment]
Thanks for the review, @jacobmontiel. I took the byte size of the serialized model (made available by MOA in the log file), maybe I should have taken the RAM/hour metric instead. I'm going to merge the PR :D
